### ISSUE TITLE
Redis grain directory key not being unregistered when target silo is known to be dead

### ISSUE THREAD

**Issue by bruno-goncalves-cko** [dotnet/orleans#7054]
Hi,

We're using MS Orleans 3.4.1 to setup a multi silo Orleans cluster within AWS with the following characteristics:

- 6 silos running under ECS fargate tasks
- DynamoDB used for grain, reminder and cluster info persistence
- Microsoft.Orleans.GrainDirectory.Redis v3.4.1-beta1 used for the grain directory
- Redis cluster with 4 nodes

We have recently observed the following error (**Target silo is known to be dead**) while a grain was trying to be activated by the cluster:

```
No target activation "S10.109.3.136:11111:355758997*cli/1db42ab3@965a24ba" for response message: "Transient Rejection (info: Target silo is known to be dead) 
Response S10.109.1.110:11111:354843009*grn/Namespace.Ommited/0+grain_id_ommited@8d48169f->S10.109.3.136:11111:355758997*cli/1db42ab3@965a24ba #1349313"
```

After inspecting the redis directory key for the affected grain I have verified the following:

 - the redis record was indeed pointing to a silo (S10.109.1.110) that was no longer running 
```
\"GrainId\":\"00000000000000000000000000000000060000005d645190+grain_id_ommited\",\"ActivationId\":\"48e8f9a7c51701350cb946358d48169f0000000000000000\",\"SiloAddress\":\"10.109.1.110:11111@354843009\"}"
```
  - the silo was indeed marked as Dead (i.e. Status = 6) on the cluster info storage
 
We often perform rolling deployments of new versions of the cluster and we expect silos to be stopped and marked as dead by Orleans while new instances of the service spin up.

From looking at the [CachedGrainLocator implementation](https://github.com/dotnet/orleans/blob/61c5a23cfe97e9f48261850446a6ce608868f27d/src/Orleans.Runtime/GrainDirectory/CachedGrainLocator.cs#L69) I would expect for the redis key to be unregistered during lookup in case the directory record still points to a Silo which has since been stopped. However, the error above seems to indicate that this is not happening.

**Restarting the cluster did not solve the issue**. Trying to activate the grain multiple times still resulted in the same error.

To resolve the issue I had to manually delete the redis key for this grain.

Would anyone please be able to shed some light on this? Unfortunately I haven't been able to reproduce this locally.

Thanks very much, please let me know if you would like me to share some more info on the matter.

**Comment by jdahl** [comment]
Could the cause for this be that the code checks against a 'knownDeadSilos' list. Not a "know alive". If the cluster is brought down isn't the membership tables cleaned if they are old enough so you have no way to know what silos where dead before?

**Comment by ReubenBond** [comment]
cc @benjaminpetit

**Comment by benjaminpetit** [comment]
> Could the cause for this be that the code checks against a 'knownDeadSilos' list. Not a "know alive". If the cluster is brought down isn't the membership tables cleaned if they are old enough so you have no way to know what silos where dead before?

I think what you are describing is possible. What settings are you using for directory cleaning? 

But we are not using a "known alive" here to try to minimize race condition: it's possible that the silo joined recently, and it's not yet propagated to the snapshot used by the redis directory.

**Comment by bruno-goncalves-cko** [comment]
> I think what you are describing is possible. What settings are you using for directory cleaning?

@benjaminpetit are you referring to the `RedisGrainDirectoryOptions`, more specifically the [`EntryExpire` option](https://github.com/dotnet/orleans/blob/61c5a23cfe97e9f48261850446a6ce608868f27d/src/Orleans.GrainDirectory.Redis/Options/RedisGrainDirectoryOptions.cs#L23)? If so, we're not setting it to any value, i.e. we leave it as the default null value as recommended by the property's comment. I assume a null value means the redis records will never expire.

If you're referring to something else, what other directory cleaning settings are there? 

As far as I can see, the dynamo table used for the cluster membership information (i.e. silos, their status and their ip addresses) does not have any cleanup/expiry options. Also, as described above, I have confirmed at the time of the issue and the cluster dynamo table did contain a document for the dead silo, as such it should have been loaded into the `KnownDeadSilos` collection on the `CachedGrainLocator` (we event restarted the cluster).

**Comment by bruno-goncalves-cko** [comment]
We have 64 documents on the cluster membership dynamo table. 6 of which have `Status = 3` and 58 have `Status = 6`.

Curiously enough we have just now experienced a similar issue where the grain wasn't being activated due to the silo host being unreacheable.

This time however (contrary to last time) the silo which was trying to be reached was indeed not on the membership table. Not sure how that happened given we don't delete anything from this table.

The error this time was:
```
No target activation "S10.109.1.60:11111:357231635*cli/9f3a42ab@87abc9be" for response message: "Transient Rejection (info: Exception while sending message: Orleans.Runtime.Messaging.ConnectionFailedException: Unable to connect to endpoint S10.109.2.19:11111:355920675. See InnerException
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Orleans.Internal.OrleansTaskExtentions.MakeCancellable[T](Task`1 task, CancellationToken cancellationToken)
   at Orleans.Runtime.Messaging.ConnectionManager.ConnectAsync(SiloAddress address)
   --- End of inner exception stack trace ---
   at Orleans.Runtime.Messaging.ConnectionManager.ConnectAsync(SiloAddress address)
   at Orleans.Runtime.Messaging.ConnectionManager.GetConnectionAsync(SiloAddress endpoint)
   at Orleans.Runtime.Messaging.OutboundMessageQueue.<SendMessage>g__SendAsync|10_0(ValueTask`1 c, Message m)) Response S10.109.2.19:11111:355920675*grn/Namespace.Omitted/0+grain_id_omitted->S10.109.1.60:11111:357231635*cli/9f3a42ab@87abc9be #65528"
```

**Comment by benjaminpetit** [comment]
> Would you be able to point me to the code where the cleanup/expiry option for the cluster membership provider is defined and configured?

It's in `ClusterMembershipOptions.DefunctSiloCleanupPeriod`. 

> Curiously enough we have just now experienced a similar issue where the grain wasn't being activated due to the silo host being unreacheable (10.109.2.19).

Are you sure only your cluster is using this Redis instance?

**Comment by bruno-goncalves-cko** [comment]
> Are you sure only your cluster is using this Redis instance?

The redis instances are being used for another purpose, but the process that's using it is not touching any of the keys used for the grain directory.

**Comment by benjaminpetit** [comment]
Only the silo that created the activation will register this entry in the directory. That means that `10.109.2.19:11111:355920675` was alive at some point.

Are you running multiple Orleans cluster?

**Comment by bruno-goncalves-cko** [comment]
> Are you running multiple Orleans cluster?

We're only running a single cluster at the moment. 

However, at one point the cluster name was different than what it is now, but at that time we did not have any grains being activated at all (the cluster was being initialized but it was behind a feature flag which prevented grains from being activated).

The above means that at the moment our cluster membership table has some silo documents referencing the old cluster name (`DeploymentId` column). From the 58 dead silos (i.e. `Status = 6`) only 36 pertain to the cluster that is currently running.

**Comment by jdahl** [comment]
The cause could be that this is run [MembershipTableCleanupAgent](https://github.com/dotnet/orleans/blob/main/src/Orleans.Runtime/MembershipService/MembershipTableCleanupAgent.cs) 
Wouldn't that clean the table of dead silos and if a silo is restarted afterwards then it wouldn't know about the dead silos and thus see the redis grain entries as alive. 
To test this you could shutdown the cluster. Clean the membership table of any provider and start it back up. If my theory holds then all redis entries are stuck.

**Comment by bruno-goncalves-cko** [comment]
@jdahl I don't think you're wrong, but I think your theory only applies if the `DefunctSiloCleanupPeriod` has been explicitly set. We're not setting it up. As such I don't think the `MembershipTableCleanupAgent` would be doing anything.

**Comment by bruno-goncalves-cko** [comment]
Well, it seems it does indeed have a non null default value: https://github.com/dotnet/orleans/blob/10af0f4af588cd4aa45cb3e250dfbffa389d59c7/src/Orleans.Core/Configuration/Options/ClusterMembershipOptions.cs#L100

**Comment by benjaminpetit** [comment]
Indeed it could explain the second situation, but not the first.

It the silo wasn't in the membership table, you would not get "Target silo is known to be dead" message

**Comment by bruno-goncalves-cko** [comment]
A similar issue has been raised [here](https://github.com/dotnet/orleans/issues/7066). @benjaminpetit there you mention a possible fix on 3.5.0. Is there an ETA for 3.5.0 to be released?

Just trying to manage any potential temporary solutions on our side

**Comment by benjaminpetit** [comment]
Hopefully next week

**Comment by ReubenBond** [comment]
Closing this as it was fixed in #7133
